### PR TITLE
Reworking mongoose transport

### PR DESCRIPTION
**Pull request by DenysGonchar**
This PR includes the following changes:
* removal of WebSocket connection support for XMPP components (non-standard and not described behaviour)
* removal of obsolete socket_type/0 callback from the mongoose_listener behaviour
* replacement of ejabberd_socket and ejabberd_receiver modules with mongoose_transport implementation
* introduction of spiral metrics for s2s and component connections
* code clean-up of the related modules
* documentation of the recently introduced metrics

### CODE REVIEW

**Comment by mongoose-im** [comment]
[small_tests_24](https://circleci.com/gh/esl/MongooseIM/168514) / small_tests / 9812e8890ef3d415e01b8b735b379d9544993926
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/3982/168514/small_tests.24.3.4) / [small](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/3982/168514/small_tests.24.3.4/small/ct_run.mongooseim@localhost.2023-03-08_22.19.29/lib.mongooseim.logs/run.2023-03-08_22.19.29/suite.log.html)

**Comment by mongoose-im** [comment]
[small_tests_24](https://circleci.com/gh/esl/MongooseIM/168540) / small_tests / 6ac2c483e0fb2ef10dcab2904e1a4fbbc320fc62
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/3982/168540/small_tests.24.3.4) / [small](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/3982/168540/small_tests.24.3.4/small/ct_run.mongooseim@localhost.2023-03-08_22.57.42/lib.mongooseim.logs/run.2023-03-08_22.57.43/suite.log.html)

**Comment by mongoose-im** [comment]
[small_tests_24](https://circleci.com/gh/esl/MongooseIM/168562) / small_tests / ba87b610ffd7218f670af1868112ef7fd265a449
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/3982/168562/small_tests.24.3.4) / [small](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/3982/168562/small_tests.24.3.4/small/ct_run.mongooseim@localhost.2023-03-09_00.08.02/lib.mongooseim.logs/run.2023-03-09_00.08.02/suite.log.html)

**Comment by mongoose-im** [comment]
[small_tests_24](https://circleci.com/gh/esl/MongooseIM/168600) / small_tests / 758c9bce79bdacb38525499efec0122e1abe4aad
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/3982/168600/small_tests.24.3.4) / [small](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/3982/168600/small_tests.24.3.4/small/ct_run.mongooseim@localhost.2023-03-09_01.05.24/lib.mongooseim.logs/run.2023-03-09_01.05.25/suite.log.html)

**Review by chrzaszcz — APPROVED** [review]
After a quick look:
- Missing migration guide information about each changed/added/removed metric. Explain reasons for the changed metrics.
- Missing information about the websockets support for components. It should be explicit in the migration guide that options are removed, especially the options enabled by default. Motivation for the removal is necessary. It would also help the reviewer understand the reason for the removal.

**Review by chrzaszcz — CHANGES_REQUESTED** [review]
Github approved for me...

**Review by NelsonVides — COMMENTED** [review]
I have two minor comments, one of them is just redundant whitespace, but the other scratches a bit on the idea of the websocket rework, so probably it can be done elsewhere and not here.
Apart from that, absolutely awesome PR 😄

**Comment by mongoose-im** [comment]
[small_tests_24](https://circleci.com/gh/esl/MongooseIM/169049) / small_tests / 59d982653ee34cb957e760ad7c723fb828d02370
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/3982/169049/small_tests.24.3.4) / [small](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/3982/169049/small_tests.24.3.4/small/ct_run.mongooseim@localhost.2023-03-10_11.48.01/lib.mongooseim.logs/run.2023-03-10_11.48.02/suite.log.html)

**Comment by mongoose-im** [comment]
[small_tests_24](https://circleci.com/gh/esl/MongooseIM/169076) / small_tests / 141cf58f789759c300cd43dce7696923e0d0facb
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/3982/169076/small_tests.24.3.4) / [small](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/3982/169076/small_tests.24.3.4/small/ct_run.mongooseim@localhost.2023-03-10_11.51.36/lib.mongooseim.logs/run.2023-03-10_11.51.37/suite.log.html)

File: doc/migrations/6.0.0_6.1.0.md [hunk 1]
```diff
@@ -7,6 +7,8 @@ With the new implementation of the client-to-server (C2S) state machine, `mongoo
 * The default value of the [`backlog`](../../configuration/listen#listenbacklog) option for all XMPP listeners has been increased from 100 to 1024 for performance reasons.
 * You might be interested in the new C2S listener options: [`max_connections`](../../configuration/listen#listenc2smax_connections), [`c2s_state_timeout`](../../configuration/listen#listenc2sc2s_state_timeout), [`reuse_port`](../../configuration/listen#listenc2sreuse_port) and [`backwards_compatible_session`](../../configuration/listen#listenc2sbackwards_compatible_session). The first two options can be set for [websockets](../../configuration/listen#handler-types-websockets-mod_websockets) as well.
 
+Support of [listen.http.handlers.mod_websockets.service] has been removed, the component connection over WebSockets did not correspond to any XEP/RFC. Also, it was never properly described in the MIM documentation.
```
Comment (NelsonVides) [inline doc/migrations/6.0.0_6.1.0.md:10]:
```suggestion
Support for [listen.http.handlers.mod_websockets.service] has been removed, the component connection over WebSockets did not correspond to any XEP/RFC, and neither it was properly described anywhere in the MIM documentation.
```
Mostly my preferred style, but no strong opinion really, just thinking out loud.

**Review by NelsonVides — COMMENTED** [review]
Quick doc changes, one is important, the other is only a suggestion :)

**Comment by DenysGonchar** [comment]
> Oh, also, why does codecov reports so bad coverage for the newly extended `mongoose_transport` module? 🤔

taking into account that it's just patch report failed, I would say that I have changed a lot of uncovered codelines.

File: src/mod_websockets.erl [hunk 2]
```diff
@@ -202,7 +161,7 @@ handle_text(Text, #ws_state{ parser = undefined } = State) ->
 handle_text(Text, #ws_state{parser = Parser} = State) ->
     case exml_stream:parse(Parser, Text) of
     {ok, NewParser, Elements} ->
-        mongoose_metrics:update(global, [data, xmpp, received, c2s, websocket, raw], byte_size(Text)),
+        mongoose_metrics:update(global, [data, xmpp, received, c2s, websocket], byte_size(Text)),
         State1 = State#ws_state{ parser = NewParser },
         case maybe_start_fsm(Elements, State1) of
```
[the remark this answers is not in the document]
Comment (DenysGonchar) [reply src/mod_websockets.erl:166]: let's make it in a context of another task. this task is about `mongoose_transport` rework and `mod_websocket` is changed only to break it's dependency with mongoose_transport.

**Comment by mongoose-im** [comment]
[small_tests_24](https://circleci.com/gh/esl/MongooseIM/169138) / small_tests / 97e3c2b039418cc84f4aa5157d94a759022150de
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/3982/169138/small_tests.24.3.4) / [small](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/3982/169138/small_tests.24.3.4/small/ct_run.mongooseim@localhost.2023-03-10_15.06.44/lib.mongooseim.logs/run.2023-03-10_15.06.45/suite.log.html)

File: src/mongoose_transport.erl [hunk 3]
```diff
@@ -1,65 +1,482 @@
...
+                                            [Socket, Shaper, Opts]),
+    Receiver.
+
+get_transport_info(ConnectionList) when is_list(ConnectionList) ->
+    [get_transport_info(Pid) || {_, Pid, _, _} <- ConnectionList, is_pid(Pid)];
+get_transport_info(TransportPid) when is_pid(TransportPid) ->
+    State = sys:get_state(TransportPid),
+    maps:from_list(lists:zip(record_info(fields, state),tl(tuple_to_list(State)))).
+
+
```
Comment (chrzaszcz) [inline src/mongoose_transport.erl:334]: Double empty line

**Review by chrzaszcz — COMMENTED** [review]
Looks ok, nice to see the legacy modules removed.
There are some untested lines in `mongoose_transport`, but I assume they were also untested before, as most of them were moved from the legacy modules.

File: big_tests/tests/s2s_SUITE.erl [hunk 4]
```diff
@@ -149,17 +154,23 @@ timeout_waiting_for_message(Config) ->
...
-
+    %% there should be at least one in and at least one out connection
+    [_ | _] = get_s2s_connections(?dh:mim(), FedDomain, in),
+    [_ | _] = get_s2s_connections(?dh:mim(), FedDomain, out),
     ok.
 
+get_s2s_connections(RPCSpec, Domain, Type)->
+    AllS2SConnections = ?dh:rpc(RPCSpec, ejabberd_s2s, get_info_s2s_connections, [Type]),
+    % ct:pal("Node = ~p, ConnectionType = ~p~nAllS2SConnections(~p): ~p",
+    %        [maps:get(node, RPCSpec), Type, length(AllS2SConnections), AllS2SConnections]),
```
[the remark this answers is not in the document]
Comment (DenysGonchar) [reply big_tests/tests/s2s_SUITE.erl:165]: I would rather leave it for now, s2s is one of the topics for the next sprint

**Comment by mongoose-im** [comment]
[small_tests_24](https://circleci.com/gh/esl/MongooseIM/169613) / small_tests / 2a186b43acf6365ed26285e833bc627b71257565
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/3982/169613/small_tests.24.3.4) / [small](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/3982/169613/small_tests.24.3.4/small/ct_run.mongooseim@localhost.2023-03-14_18.13.45/lib.mongooseim.logs/run.2023-03-14_18.13.45/suite.log.html)

**Comment by mongoose-im** [comment]
[small_tests_24](https://circleci.com/gh/esl/MongooseIM/169636) / small_tests / c75eb3ca4a443c44e10e0ab475bfaccaa7e8551a
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/3982/169636/small_tests.24.3.4) / [small](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/3982/169636/small_tests.24.3.4/small/ct_run.mongooseim@localhost.2023-03-15_02.25.55/lib.mongooseim.logs/run.2023-03-15_02.25.55/suite.log.html)

File: doc/operation-and-maintenance/MongooseIM-metrics.md [hunk 5]
```diff
@@ -161,9 +161,22 @@ Metrics specific to an extension, e.g. Message Archive Management, are described
...
 | ----------- | ---- | ----------- |
-| `[global, data, xmpp, received, xml_stanza_size]` | histogram | A size (in bytes) of a received stanza after decompression and decryption. |
-| `[global, data, xmpp, sent, xml_stanza_size]` | histogram | A size (in bytes) of a stanza sent to a client socket. |
-| `[global, data, dist]` | proplist | Network stats for an Erlang distributed communication. A proplist with values: `recv_oct`, `recv_cnt`, `recv_max`, `send_oct`, `send_max`, `send_cnt`, `send_pend`, `connections` |
+| `[global, data, xmpp, received, xml_stanza_size]` | histogram | A size (in bytes) of a received stanza after decryption. |
+| `[global, data, xmpp, sent, xml_stanza_size]` | histogram | A size (in bytes) of a sent stanza before encryption. |
+| `[global, data, xmpp, received, c2s, tcp]` | spiral | A size (in bytes) of unencrypted data received from a client via TCP channel. |
+| `[global, data, xmpp, sent, c2s, tcp]` | spiral | A size (in bytes) of unencrypted data sent to a client via TCP channel. |
+| `[global, data, xmpp, received, c2s, tls]` | spiral | A size (in bytes) of a data received from a client via TLS channel after decryption. |
```
Comment (chrzaszcz) [inline doc/operation-and-maintenance/MongooseIM-metrics.md:168]: It's still  "a data" in all these lines.

**Review by chrzaszcz — APPROVED** [review]
Looks good 👍

**Pull request merged**